src/mono/mono/metadata/loader.c

```diff
@@ -823,7 +823,6 @@ mono_method_search_in_array_class (MonoClass *klass, const char *name, MonoMetho
 	int i;
 
 	mono_class_setup_methods (klass);
-	g_assert (!mono_class_has_failure (klass)); /*FIXME this should not fail, right?*/
```

Comment:
@vargaz I removed this assertion to fix a problem related to this feature. Do you see any potential problem with the callsites in `src/mono/mono/mini/mini-generic-sharing.c`?

Reply:
Why does this fail ? Also, if its removed, wouldn't the rest of the code fail to work ?

Reply:
so the problem is if code does this:

```
newobj  SomeRefStruct[]::.ctor(int)
```

method-to-ir ends up trying to find the .ctor for an array of a byreflike type.  that class is marked as failed.   

the problem is that `newobj` has a token for a method, so we end up calling `method_from_memberref`  on that token.   We can't tell that `SomeRefStruct[]` is a bad type because we don't know the type - we only know that there's a method token.

So if you remove `g_assert(!mono_class_had_failure(...))` we will return that ctor back to method-to-ir.  and then we can look at the `cmethod->klass` and figure out that it's an invalid program.

Reply:
what we could probably do is split up `method_from_member` ref into two functions:  one that just finds the parent class of a memberref token, and the second one that find the method given the parent class and the token.

then method-to-ir could call the two parts seprately.  given the token for `SomeRefStruct[]::.ctor(int)`, it can find the class `SomeRefStruct[]` and then fail it.

---

the other thing is that `mono_class_setup_methods()` apparently works for failed array types - it just goes and creates the constructors (and the generic impl methods?).  that's why removing this assert seems to fix the test.

The question is what happens to other callers of `mono_method_search_in_array_class` - there are two in generic sharing.  We weren't sure how they're expected to behave.

Reply:
The question is whenever this can fail at runtime, i.e. whenever generic sharing can encounter a type where these calls fail. In the meantime, we should at least assert at the call sites.

Reply:
for generic sharing case, I think we should assert or throw an exception if `klass` has failure before calling `mono_method_search_in_array_class`. `mono_method_search_in_array_class` should assume being given a valid class to look for methods for. The callsite of `mono_method_search_in_array_class` should have more context in terms of why the class would have failure.